smolyay/basis.py

```diff
@@ -1,29 +1,33 @@
 import abc
-import math
 
 import numpy
 import scipy.special
 
+
 class BasisFunction(abc.ABC):
     """Basis function for interpolating data.
 
-     A one-dimensional basis function is defined on the domain
-     :math:`[-1,1]`. The function defines the :attr:`points` at
-     which it should be sampled within this interval for interpolation.
-     The function also has an associated :meth:`__call__` method
-     for evaluating it at a point within its domain. Moreover,
-     the first derivative of the function can be evaluated via
-     :meth:`derivative`.
+    A one-dimensional basis function is defined on some given natural 
+    domain. The function defines the :attr:`points` at which it should 
+    be sampled within this interval for interpolation. The function also 
+    has an associated :meth:`__call__` method for evaluating it at a 
+    point within its domain. Moreover, the first derivative of the function 
+    can be evaluated via :meth:`derivative`.
+
+    Parameters
+    ----------
+    natural_domain : list of two numbers
+        the natural domain of the function within which calls are made.
     """
 
-    def __init__(self):
+    def __init__(self, natural_domain):
         self._points = []
+        self._natural_domain = natural_domain
 
     @property
-    @abc.abstractmethod
-    def points(self):
-        """list: Sampling points for interpolation."""
-        pass
+    def natural_domain(self):
+        """list: Domain the sample points come from."""
+        return self._natural_domain
 
     @abc.abstractmethod
     def __call__(self, x):
```

Comment:
In the benchmark base class, it had the `__call__` method that would take care of sizing issues as well as check if the inputs were in the domain, and then it would return the answer of an abstract `_function` class. Do you think we should do something similar here?

And on that note, the input of the surrogate `__call__ ` and `train_from_data` also have sizing issues to take care of in addition to normalization. Should taking care of the sizing, normalization, and detecting invalid inputs be the job of the base class?

Reply:
Hmm, good thought! Yes, I think I would do a concrete `__call__` (and `derivative`) method in the base class, and have `_function` (and `_derivative`) abstract methods that deriving classes must override.

In the `__call__` method,  we can make a decision whether to throw an error if the inputs are not in the right domain or if it is a contract with the caller that the inputs are valid. There is a performance hit to error checking, but the error messages downstream will be easier for the caller to understand. I am good either way.

I don't think we should do any casting or reshaping of `x` in `__call__` before forwarding to `_function` because the `_function` should make a decision about whether that is needed (e.g., the Chebyshev function will just call scipy).

We do expect that `__call__` will return an output consistent with the shape of `x` (a scalar if `x` is a number, otherwise a NumPy array with the matching shape). This logic should be included here.